modules/healthcheck/default.nix

```
@@ -0,0 +1,281 @@
+                  }
+                );
+              }
+            ) (servicesWithHealthcheck);
+
+            services =
+              (lib.mapAttrs (
+                mainServiceName: serviceConfig:
+                let
+                  cfg = serviceConfig.healthcheck;
+                in
+                (mkIf (cfg.readiness-probe.enable) (
+                  let
+                    probeCfg = cfg.readiness-probe;
+                  in
+                  {
+                    serviceConfig.Type = lib.mkForce "notify";
+                    serviceConfig.TimeoutStartSec = lib.mkForce "infinity";
+
+                    # 2. Build the wrapper script that becomes the new ExecStart.
+                    serviceConfig.ExecStart =
+                      let
+                        scriptPath = makeBinPath (
+                          [
+                            pkgs.systemd
+                            pkgs.curl
+                            pkgs.gawk
+                          ]
+                          ++ (if cfg ? "path" then cfg.path else [ ])
+                          ++ (if serviceConfig ? "path" then serviceConfig.path else [ ])
+                        );
+                        successStatus = builtins.replaceStrings [ "%OUTPUT%" ] [ ''''${output}'' ] probeCfg.statusReady;
+                      in
+                      lib.mkForce (
+                        pkgs.writeShellScript "${mainServiceName}-readiness-check" ''
+                          #!${pkgs.runtimeShell}
+                          # set -o errexit
+                          set -o nounset
+                          export PATH="${scriptPath}:$PATH"
+
+                          check() {
+                            echo "Health check: starting background readiness probe for ${mainServiceName}."
+                            while true; do
+                              local output
+                              output=$(sh -c "${probeCfg.command}" 2>/dev/null)
+
+                              if [[ -n "''${output}" ]]; then
+                                echo "Health check: probe successful. Notifying systemd that service is ready."
+                                systemd-notify --ready --status="${successStatus}"
+                                return 0
+                              else
+                                echo "Health check: probe not successful. Notifying systemd that service is still waiting."
+                                systemd-notify --status="${probeCfg.statusWaiting}"
+                              fi
+                              sleep ${toString probeCfg.periodSeconds}
+                            done
+                          }
+
+                          if [[ -n "''${NOTIFY_SOCKET:-}" ]]; then
+                            check &
+                          else
+                            echo "Health check: NOTIFY_SOCKET not set. Cannot run readiness probe." >&2
```

This is a fatal problem, so it's best to exit.

```suggestion
                            echo "Health check: NOTIFY_SOCKET not set. Cannot run readiness probe." >&2
                            exit 1
```